src/persist-client/src/impl/encoding.rs

```diff
@@ -0,0 +1,416 @@
+    }
+
+    fn from_proto(proto: String) -> Result<Self, TryFromProtoError> {
+        match parse_id('r', "ReaderId", &proto) {
+            Ok(x) => Ok(ReaderId(x)),
+            Err(_) => Err(TryFromProtoError::InvalidShardId(proto)),
+        }
+    }
+}
+
+impl<K, V, T, D> State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    pub fn decode(buf: &[u8]) -> Result<Self, InvalidUsage<T>> {
+        let proto = ProtoStateRollup::decode_checksummed(buf)
+            // We received a State that we couldn't decode. This could happen if
+            // persist messes up backward/forward compatibility, if the durable
+            // data was corrupted, or if operations messes up deployment. In any
+            // case, fail loudly.
+            .expect("internal error: invalid encoded state");
+        Self::try_from(proto).expect("internal error: invalid encoded state")
+    }
+}
+
+impl<K, V, T, D> Codec for State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    fn codec_name() -> String {
+        "md5+protobuf[State]".into()
+    }
+
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+    {
+        // TODO: Move checksum to be a field on the proto instead. We can encode
+        // the proto, checksum'ing as we go, and then manually append it onto
+        // the end.
+        //
+        // TODO: Regardless of the above TODO, compute the checksum as we go and
+        // avoid this temp Vec.
+        let temp = self.into_proto().encode_to_vec();
+        buf.put_slice(&temp);
+        let checksum = ProtoStateRollup::md5_checksum(&temp);
+        buf.put_slice(&checksum);
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        let proto = ProtoStateRollup::decode_checksummed(buf)?;
+        // This match goes away when we do incremental state.
+        match State::try_from(proto) {
+            Ok(Ok(x)) => Ok(x),
+            Ok(Err(err)) => Err(err.to_string()),
+            Err(err) => Err(err.to_string()),
+        }
+    }
+}
+
+impl ProtoStateRollup {
+    // NB: This len is intentionally hardcoded (not derived from the md5 crate)
+    // so that a change to the crate can't break us. The compiler statically
+    // checks that they match because this const is used in the return type.
+    // (MD5 is not going to change, so this is all a moot point, but still
+    // better to be defensive.)
+    const CHECKSUM_LEN: usize = 16;
+
+    fn decode_checksummed(buf: &[u8]) -> Result<ProtoStateRollup, String> {
+        if buf.len() < Self::CHECKSUM_LEN {
+            return Err("missing/incomplete checksum".into());
+        }
+        let checksum_start = buf.len() - Self::CHECKSUM_LEN;
+        let (buf, checksum) = (&buf[..checksum_start], &buf[checksum_start..]);
+        if checksum != &Self::md5_checksum(&buf) {
+            return Err("checksum mismatch".into());
```

Comment:
will the overall stack give us enough information to know where to look if we ever hit this?